File: typescript/base/src/main/vsdk/toolkit/processing/CurveModeler.ts
// Vitral classes
import { Vector3Dd } from "../common/linealAlgebra/Vector3Dd.js";
import { ParametricCurve } from "../environment/geometry/curve/ParametricCurve.js";
import { ProcessingElement } from "./ProcessingElement.js";

/**
Utility class with static geometry algorithms, mostly for creating and
modifying geometric entities.

It complements:
- `ComputationalGeometry` (geometric queries)
- `SimpleTestGeometryLibrary` (sample geometry generation)

This class does not render anything. It only manipulates geometric data
structures and delegates visualization concerns to renderers.
*/
export class CurveModeler extends ProcessingElement {
    /**
    Creates a 3D line from point (x1, y1, z1) to point (x2, y2, z2).
    */
    public static createLine(x1: number, y1: number, z1: number, x2: number, y2: number, z2: number): ParametricCurve {
        let lineModel: ParametricCurve;
        let pointParameters: Vector3Dd[];

        lineModel = new ParametricCurve();
        pointParameters = new Array<Vector3Dd>(1);
        pointParameters[0] = new Vector3Dd(x1, y1, z1);
        lineModel.addPoint(pointParameters, ParametricCurve.CORNER);

        pointParameters = new Array<Vector3Dd>(1);
        pointParameters[0] = new Vector3Dd(x2, y2, z2);
        lineModel.addPoint(pointParameters, ParametricCurve.CORNER);

        return lineModel;
    }
}

File: cpp/base/src/main/vsdk/toolkit/processing/CurveModeler.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/geometry/curve/ParametricCurve.h"
#include "vsdk/toolkit/processing/CurveModeler.h"

ParametricCurve* CurveModeler::createLine(double x1, double y1, double z1,
                                          double x2, double y2, double z2)
{
    ParametricCurve* lineModel;
    java::ArrayList<Vector3Dd> pointParameters;

    lineModel = new ParametricCurve();
    pointParameters.add(Vector3Dd(x1, y1, z1));
    lineModel->addPoint(pointParameters, ParametricCurve::CORNER);

    pointParameters.clear();
    pointParameters.add(Vector3Dd(x2, y2, z2));
    lineModel->addPoint(pointParameters, ParametricCurve::CORNER);

    return lineModel;
}
